Stride-Labs/stride
# Autopilot enhancement

## Context and purpose of the change

The goal of this PR is to support two more features on autopilot module

* RedeemStake as part of ibc transfer
* LiquidStake and IBC transfer back liquid staking token as part of ibc transfer

Note: This PR moved from https://github.com/Stride-Labs/stride/pull/626

## Brief Changelog

- MsgLiquidStake is returning received liquid staking token amount
- Add unit test for RedeemStake and LiquidStakeAndIBCTransfer
- Add integration test for RedeemStake and LiquidStakeAndIBCTransfer

## Author's Checklist

I have...

- [x] Run and PASSED locally all GAIA integration tests
- [x] If the change is contentful, I either:
    - [x] Added a new unit test OR 
    - [x] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.


## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

### How to test integration test on local

- Update `DefaultStakeibcActive` to `true` at `x/autopilot/types/params.go`
- Comment out all `skip "DefaultActive set to false, skip test"` on `dockernet/tests/integration_tests.bats`
- Install docker image for stride `make build-docker build=s`
- Start dockernet with `make start-docker`
- Run integration test with `make test-integration-docker`
- Ensure the result is as following
```
integration_tests.bats
 ✓ [INTEGRATION-BASIC-GAIA] host zones successfully registered
 ✓ [INTEGRATION-BASIC-GAIA] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-GAIA] liquid stake mint and transfer
 ✓ [INTEGRATION-BASIC-GAIA] packet forwarding automatically liquid stakes
 ✓ [INTEGRATION-BASIC-GAIA] transfer stuatom to host chain
 ✓ [INTEGRATION-BASIC-GAIA] packet forwarding automatically liquid stake and ibc transfer stAsset to original network
 ✓ [INTEGRATION-BASIC-GAIA] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC-GAIA] packet forwarding automatically redeem stake
 ✓ [INTEGRATION-BASIC-GAIA] redemption works
 ✓ [INTEGRATION-BASIC-GAIA] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-GAIA] rewards are being reinvested, exchange rate updating
 ✓ [INTEGRATION-BASIC-GAIA] rewards are being distributed to stakers

12 tests, 0 failures
```

### How to test Hub -> Stride -> Osmosis

- Update `dockernet/config.sh` to `HOST_CHAINS=(GAIA OSMO)`
- Start dockernet with `make start-docker build=sgor`
- Test following script and ensure OSMO_ADDRESS gets stuatom ibc denom 
```
GAIA_ADDRESS=$(./build/gaiad keys show -a gval1 --home=./dockernet/state/gaia1)
STRIDE_ADDRESS=stride1jrmtt5c6z8h5yrrwml488qnm7p3vxrrml2kgvl
OSMO_ADDRESS=osmo1dtq0y9reqst7d99fd3c7x6dflh4eazm4lx5sk9
GAIA_STRIDE_CHAN=channel-0
STRIDE_OSMO_CHAN=channel-1
memo='{ "autopilot": { "receiver": "'$STRIDE_ADDRESS'",  "stakeibc": { "stride_address": "'$STRIDE_ADDRESS'", "action": "LiquidStake", "ibc_receiver": "'$OSMO_ADDRESS'", "transfer_channel": "'$STRIDE_OSMO_CHAN'"  } } }'
./build/gaiad tx ibc-transfer transfer transfer $GAIA_STRIDE_CHAN "$memo" 10000uatom --from gval1 -y --home=./dockernet/state/gaia1
./build/gaiad query tx 31A1480A5F025F900A684F7CFD92F2FE06B4307CA81BD1E01AB31E9C01BFC7DF --home=./dockernet/state/gaia1
./build/gaiad query bank balances $GAIA_ADDRESS --home=./dockernet/state/gaia1
./build/strided query bank balances $STRIDE_ADDRESS --home=./dockernet/state/stride1
./build/osmosisd query bank balances $OSMO_ADDRESS --home=./dockernet/state/osmo1
```

## Documentation and Release Note

  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] not documented
